src/pandaboxlib/pandaboxlib.py

```diff
@@ -6,80 +6,282 @@
 
     def connect_to_panda(self):
-        self.sock.connect((self.host, self.port))
+        """Create socket connection to host"""
+        if self.sock is None:
+            self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            self.sock.setsockopt(
+                socket.SOL_TCP,          # Disable Nagle algorithm
+                socket.TCP_NODELAY,      # (https://linux.die.net/man/7/tcp)
+                1                        # Is this really necessary?
+            )
+            self.sock.settimeout(1)
+            self.sock.connect((self.host, self.port))
 
     def disconnect_from_panda(self):
-        self.sock.shutdown(socket.SHUT_WR)
-        self.sock.close()
+        """Close socket connection to host"""
+        if self.sock is not None:
+            self.sock.shutdown(socket.SHUT_WR)
+            self.sock.close()
+            self.sock = None
 
     def query(self, cmd):
+        """Send command to host and return response"""
         self.sock.sendall((cmd + '\n').encode())
         val = str(self.sock.recv(4096).decode())
         print(val)
         return val
 
     def query_value(self, cmd):
+        """Send command to host and return response as float"""
         val = self.query(cmd)
         return float(val.strip("\n").split("=")[1])
 
     def _num(self, s):
+        """Cast value to int, else float"""
         try:
             return int(s)
         except ValueError:
             return float(s)
 
     def numquery(self, cmd):
+        """Send command to host and return response as int or float"""
         val = self.query(cmd)
         val = val.split("=")[-1]
         val = self._num(val)
         print(str(val))
         return val
 
-    def save_config(self, save_file):
-        input = GetLines(self.sock)
-        output = open(save_file, 'w')
-
-        # First save the CONFIG state
-        save_state(input, output, '*CHANGES.ATTR?', self.sock)
-        save_state(input, output, '*CHANGES.CONFIG?', self.sock)
-        # Now save the individual tables.
-        # Note that we must read the complete table response before processing!
-        tables = list(read_response(input, '*CHANGES.TABLE?', self.sock))
-        for table in tables:
-            save_table(input, output, table, self.sock)
-
-        # Finally the metadata is a bit more tricky, because the table and value results
-        # are mixed together.
-        metadata = list(read_response(input, '*CHANGES.METADATA?', self.sock))
-        for line in metadata:
-            save_metadata(input, output, line, self.sock)
-
-    def load_config(self, load_file):
-        for line in open(load_file):
-            self.sock.sendall(line.encode())
+    def save_config(self, path):
+        """Save design to file"""
+        if self.sock is not None:
+            self.disconnect_from_panda()
+        self.connect_to_panda()              # Ensure first `*CHANGES` request on connection
+        _Design(self.sock).save(path)
+
```

Comment:
Can you please add a try-except around this, in which there will be a message giving some context, such as "unable to save design to path %s"?

Reply:
Will do!